components/director/internal/systemfetcher/template.go

```diff
@@ -0,0 +1,159 @@
+			Name: p.PlaceholderName,
+		})
+	}
+
+	return &renderer{
+		appTemplateService:   appTemplateService,
+		appConverter:         appConverter,
+		appInputOverride:     appInputOverride,
+		placeholdersMapping:  mapping,
+		placeholdersOverride: placeholders,
+	}
+}
+
+func (r *renderer) ApplicationRegisterInputFromTemplate(ctx context.Context, sc System) (*model.ApplicationRegisterInput, error) {
+	appTemplate, err := r.appTemplateService.Get(ctx, sc.TemplateID)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while getting application template with ID %s", sc.TemplateID)
+	}
+
+	inputValues, err := r.getTemplateInputs(sc)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while getting template inputs for Application Template with name %s", appTemplate.Name)
+	}
+
+	appTemplate.Placeholders = r.placeholdersOverride
+	appTemplate.ApplicationInputJSON, err = r.mergedApplicationInput(appTemplate.ApplicationInputJSON, r.appInputOverride)
+	if err != nil {
+		return nil, errors.Wrap(err, "while merging application input from template and override application input")
+	}
+	appRegisterInputJSON, err := r.appTemplateService.PrepareApplicationCreateInputJSON(appTemplate, *inputValues)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while preparing ApplicationRegisterInput JSON from Application Template with name %s", appTemplate.Name)
+	}
+
+	appRegisterInput, err := r.appConverter.CreateInputJSONToModel(ctx, appRegisterInputJSON)
+	if err != nil {
+		return nil, errors.Wrapf(err, "while preparing ApplicationRegisterInput model from Application Template with name %s", appTemplate.Name)
+	}
+
+	return &appRegisterInput, nil
+}
+
+func (r *renderer) getTemplateInputs(s System) (*model.ApplicationFromTemplateInputValues, error) {
+	systemJSON, err := json.Marshal(s)
+	if err != nil {
+		return nil, err
+	}
+
+	inputValues := model.ApplicationFromTemplateInputValues{}
+	for _, pm := range r.placeholdersMapping {
+		placeholderInput := gjson.GetBytes(systemJSON, pm.SystemKey).String()
+		inputValues = append(inputValues, &model.ApplicationTemplateValueInput{
+			Placeholder: pm.PlaceholderName,
+			Value:       placeholderInput,
+		})
+	}
+
+	return &inputValues, nil
+}
+
+func (r *renderer) mergedApplicationInput(originalAppInputJSON, overrideAppInputJSON string) (string, error) {
+	var originalAppInput map[string]interface{}
+	var overrideAppInput map[string]interface{}
+
+	if err := json.Unmarshal([]byte(originalAppInputJSON), &originalAppInput); err != nil {
+		return "", errors.Wrapf(err, "while unmarshaling original application input")
+	}
+
+	if err := json.Unmarshal([]byte(overrideAppInputJSON), &overrideAppInput); err != nil {
+		return "", errors.Wrapf(err, "while unmarshaling override application input")
+	}
+
+	for field, overrideV := range overrideAppInput {
+		if originalV, ok := originalAppInput[field]; ok && originalV != nil {
+			if reflect.TypeOf(originalV) != reflect.TypeOf(overrideV) {
```

Comment:
What does this checks for? Isn't always strings? Or one can be a string and another object or array?